neuraxle/metaopt/tpe.py

```diff
@@ -0,0 +1,293 @@
+            sorted_stds[1:-1] = np.maximum(sorted_means[1:-1] - sorted_means[0:-2],
+                                           sorted_means[2:] - sorted_means[1:-1])
+            left_std = sorted_means[1] - sorted_means[0]
+            right_std = sorted_means[-1] - sorted_means[-2]
+            sorted_stds[0] = left_std
+            sorted_stds[-1] = right_std
+
+        # Magic formula from hyperopt.
+        # -- magic formula:
+        # maxsigma = old_div(prior_sigma, 1.0)
+        # minsigma = old_div(prior_sigma, min(100.0, (1.0 + len(srtd_mus))))
+        #
+        # sigma = np.clip(sigma, minsigma, maxsigma)
+        # sigma[prior_pos] = prior_sigma
+        min_std = prior_sigma / min(100.0, (1.0 + len(sorted_means)))
+        max_std = prior_sigma / 1.0
+        sorted_stds = np.clip(sorted_stds, min_std, max_std)
+
+        if self.use_linear_forgetting_weights:
+            distribution_amplitudes = _linear_forgetting_Weights(len(means), self.number_recent_trial_at_full_weights)
+        else:
+            # From tpe article : TPE substitutes an equally-weighted mixture of that prior with Gaussians centered at each observations.
+            distribution_amplitudes = np.ones(len(means))
+            # distribution_amplitudes = [hyperparam_distribution.pdf(mean) for mean in means]
+
+        if use_prior:
+            sorted_stds[prior_pos] = prior_sigma
+            sorted_distribution_amplitudes = np.zeros_like(sorted_means)
+            sorted_distribution_amplitudes[:prior_pos] = distribution_amplitudes[sort_indexes[:prior_pos]]
+            sorted_distribution_amplitudes[prior_pos] = sort_indexes
+            sorted_distribution_amplitudes[prior_pos + 1:] = distribution_amplitudes[sort_indexes[prior_pos:]]
+        else:
+            sorted_distribution_amplitudes = distribution_amplitudes
+
+        # Normalize distribution amplitudes.
+        distribution_amplitudes = np.array(distribution_amplitudes)
+        distribution_amplitudes /= np.sum(distribution_amplitudes)
+
+        return sorted_distribution_amplitudes, sorted_means, sorted_stds, distributions_mins, distributions_max
+
+    def find_next_best_hyperparams(self, auto_ml_container: 'AutoMLContainer') -> HyperparameterSamples:
+        """
+        Find the next best hyperparams using previous trials.
+
+        :param auto_ml_container: trials data container
+        :type auto_ml_container: Trials
+        :return: next best hyperparams
+        :rtype: HyperparameterSamples
+        """
+        # Flatten hyperparameter space
+        flat_hyperparameter_space = auto_ml_container.hyperparameter_space.to_flat()
+
+        if auto_ml_container.trial_number < self.number_of_initial_random_step:
+            # Perform random search
+            return self.initial_auto_ml_algo.find_next_best_hyperparams(auto_ml_container)
+
+        # Keep only success trials
+        success_trials = auto_ml_container.trials.filter(TRIAL_STATUS.SUCCESS)
```

Comment:
Maybe could be possible to affect an infinite score value  (or infinite negative score value), I thinks that how they do in hyperopt currently, in order to be always selected in the bad distribution. If not, could be possible to add it to the bad distribution after having split the valid trial into good and bad.
Maybe TPE is already a little bit robust against that. Since in TPE, we only select candidates in the good distribution, but if the good distribution reach over failed experiment that might be an issue. 

Might be a good idea to create unittests with some failed trial.